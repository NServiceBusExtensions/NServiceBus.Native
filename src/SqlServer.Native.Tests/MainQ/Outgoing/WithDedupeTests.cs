using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using NServiceBus.Transport.SqlServerNative;
using ObjectApproval;
using Xunit;
using Xunit.Abstractions;

public class WithDedupeTests :
    TestBase
{
    static DateTime dateTime = new DateTime(2000, 1, 1, 1, 1, 1, DateTimeKind.Utc);

    string table = "WithDedupeTests";

    [Fact]
    public async Task Single()
    {
        var message = BuildBytesMessage("00000000-0000-0000-0000-000000000001");
        var database = await LocalDb();
        using (var connection = await database.OpenConnection())
        {
            var manager = new QueueManager(table, connection, "Deduplication");
            await manager.Create();
            var dedupeManager = new DedupeManager(connection, "Deduplication");
            await dedupeManager.Create();
            await manager.Send(message);
            ObjectApprover.VerifyWithJson(await SqlHelper.ReadData(table, connection));
        }
    }

    [Fact]
    public async Task Single_WithDuplicate()
    {
        var message = BuildBytesMessage("00000000-0000-0000-0000-000000000001");
        var database = await LocalDb();
        using (var connection = await database.OpenConnection())
        {
            var manager = new QueueManager(table, connection, "Deduplication");
            await manager.Create();
            var dedupeManager = new DedupeManager(connection, "Deduplication");
            await dedupeManager.Create();
            await manager.Send(message);
            await manager.Send(message);
            ObjectApprover.VerifyWithJson(await SqlHelper.ReadData(table, connection));
        }
    }

    //[Fact]
    //public void Single_WithPurgedDuplicate()
    //{
    //    var message = BuildBytesMessage("00000000-0000-0000-0000-000000000001");
    //    Send(message);
    //    Send(message);
    //    ObjectApprover.VerifyWithJson(SqlHelper.ReadData(table));
    //}

    [Fact]
    public async Task Batch()
    {
        var messages = new List<OutgoingMessage>
        {
            BuildBytesMessage("00000000-0000-0000-0000-000000000001"),
            BuildBytesMessage("00000000-0000-0000-0000-000000000002")
        };
        var database = await LocalDb();
        using (var connection = await database.OpenConnection())
        {
            var manager = new QueueManager(table, connection, "Deduplication");
            await manager.Create();
            var dedupeManager = new DedupeManager(connection, "Deduplication");
            await dedupeManager.Create();
            await manager.Send(messages);
            ObjectApprover.VerifyWithJson(await SqlHelper.ReadData(table, connection));
        }
    }

    [Fact]
    public async Task Batch_WithFirstDuplicate()
    {
        var message = BuildBytesMessage("00000000-0000-0000-0000-000000000001");
        var messages = new List<OutgoingMessage>
        {
            BuildBytesMessage("00000000-0000-0000-0000-000000000001"),
            BuildBytesMessage("00000000-0000-0000-0000-000000000002")
        };
        var database = await LocalDb();
        using (var connection = await database.OpenConnection())
        {
            var manager = new QueueManager(table, connection, "Deduplication");
            await manager.Create();
            var dedupeManager = new DedupeManager(connection, "Deduplication");
            await dedupeManager.Create();
            await manager.Send(message);
            await manager.Send(messages);
            ObjectApprover.VerifyWithJson(await SqlHelper.ReadData(table, connection));
        }
    }

    [Fact]
    public async Task Batch_WithSecondDuplicate()
    {
        var message = BuildBytesMessage("00000000-0000-0000-0000-000000000002");
        var messages = new List<OutgoingMessage>
        {
            BuildBytesMessage("00000000-0000-0000-0000-000000000001"),
            BuildBytesMessage("00000000-0000-0000-0000-000000000002")
        };
        var database = await LocalDb();
        using (var connection = await database.OpenConnection())
        {
            var manager = new QueueManager(table, connection, "Deduplication");
            await manager.Create();
            var dedupeManager = new DedupeManager(connection, "Deduplication");
            await dedupeManager.Create();
            await manager.Send(message);
            await manager.Send(messages);
            ObjectApprover.VerifyWithJson(await SqlHelper.ReadData(table, connection));
        }
    }

    static OutgoingMessage BuildBytesMessage(string guid)
    {
        return new OutgoingMessage(new Guid(guid), dateTime, "headers", Encoding.UTF8.GetBytes("{}"));
    }

    public WithDedupeTests(ITestOutputHelper output) :
        base(output)
    {
    }
}