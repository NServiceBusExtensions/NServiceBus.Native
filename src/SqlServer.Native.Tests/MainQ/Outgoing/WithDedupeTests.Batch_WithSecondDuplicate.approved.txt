[
  {
    $type: 'IncomingVerifyTarget',
    Expires: 'DateTime 1',
    Headers: 'headers',
    Id: 'Guid 1',
    Body: '{}'
  },
  {
    $type: 'IncomingVerifyTarget',
    Expires: 'DateTime 1',
    Headers: 'headers',
    Id: 'Guid 2',
    Body: '{}'
  }
]